### PR TITLE
S3949: Add C# 9 tests and do not fail for unsupported language version

### LINKED ISSUE #3439: Update S3949 and CFG creation for C# 9 syntax
S3949 is generating MLIR instructions for all C# syntaxes, based on the CFG. When encountering an unknown syntax, it will throw an exception together with an AD0001 warning. It fails for C# 9 syntax.

```
error AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CbdeHandlerRule' threw an exception of type 'SonarAnalyzer.CBDE.CbdeException' with message 'Top level error in CBDE handling: WithExpression 
 Details:  
 Inner exception:  
 Stack trace:    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.BuildExpression(ExpressionSyntax expression, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.BuildVariableDeclarator(VariableDeclaratorSyntax variableDeclarator, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.BuildVariableDeclaration(VariableDeclarationSyntax declaration, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.BuildStatement(StatementSyntax statement, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.BuildStatements(IEnumerable`1 statements, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.BuildBlock(BlockSyntax block, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.BuildStatement(StatementSyntax statement, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraphBuilder.Build(SyntaxNode node, Block currentBlock) 
    at SonarAnalyzer.ControlFlowGraph.AbstractControlFlowGraphBuilder.Build() 
    at SonarAnalyzer.ControlFlowGraph.CSharp.CSharpControlFlowGraph.Create(CSharpSyntaxNode node, SemanticModel semanticModel) 
    at SonarAnalyzer.CBDE.MlirExporter.ExportFunction(MethodDeclarationSyntax method) 
    at SonarAnalyzer.CBDE.CbdeHandler.ExportFunctionMlir(SyntaxTree tree, SemanticModel model, MlirExporterMetrics exporterMetrics, String mlirFileName) 
    at SonarAnalyzer.CBDE.CbdeHandler.<RegisterMlirAndCbdeInOneStep>b__31_0(CompilationAnalysisContext c)'. [C:\Workspace\sonar-dotnet\sonaranalyzer-dotnet\its\sources\Net5\Net5.csproj
```

Note: The regression ITs are considering AD0001 warnings as errors. Thus when running the `Net5` integration test, we disable S3949. We should re-enable it after fixing this issue (see sonaranalyzer-dotnet/its/regression-test.ps1).

### LINKED ISSUE #3697: Add tests w/ C# 9 syntax for CBDE rule and MLIR exporter
# CBDE
- [ ] S3949	CbdeHandlerRule.cs
- [ ] MlirExporter

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #3439
Fixes #3697

I did not add tests for top level statement, because test infrastructure and implementation works on `MethodDeclarationSyntax` level.

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
I don't see a way how to make the coverage happy in this case. We don't throw from CFG for C# 8 and lower and I guess I can't attach a debugger during a test run :)

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 1]
```diff
@@ -513,11 +492,7 @@ try {
     Build-Project-MSBuild "SkipGeneratedVb" "SkipGeneratedVb.sln"
 
     Build-Project-DotnetTool "NetCore31" "NetCore31.sln"
-
-    # The CBDE rule is failing for C# 9 syntax. See https://github.com/SonarSource/sonar-dotnet/issues/3439
-    DisableRule "S3949"
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/its/regression-test.ps1:518]: ah, so this is why everything was green

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/MlirExporter.cs [hunk 2]
```diff
@@ -137,6 +139,19 @@ public void ExportFunction(MethodDeclarationSyntax method)
...
             }
+
+            IControlFlowGraph cfg;
+            try
+            {
+                cfg = CSharpControlFlowGraph.Create(method.Body, semanticModel);
+            }
+            catch (Exception ex) when (semanticModel.Compilation.GetLanguageVersion() > supportedVersion && !Debugger.IsAttached)
+            {
+                // We don't want to fail when exception occurs for unsupported language version
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/MlirExporter.cs:150]: nitpick
```suggestion
                // We don't want to fail when the exception occurs for an unsupported language version
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/MlirExporter.cs [hunk 3]
```diff
@@ -137,6 +139,19 @@ public void ExportFunction(MethodDeclarationSyntax method)
...
+                writer.WriteLine($"// Skipping function {method.Identifier.ValueText}{GetAnonymousArgumentsString(method)}, failed to generate CFG: {ex.Message}");
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/MlirExporter.cs:151]: does this get written in the MLIR?
can you add an integration test to verify this? 

I know it's a bunch of work, but I don't feel comfortable with what we currently have

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/CBDE/MlirExportTest.cs [hunk 4]
```diff
@@ -1325,6 +1325,90 @@ void f()
...
+    private List<int> list;
+
+    public void Method()
+    {
+        list = new();
+        StringBuilder sb = new();
+        StringBuilder sb2 = new(null);
+        StringBuilder sb3 = new(length: 4, capacity: 3, startIndex: 1, value: ""fooBar"");
+        Console.WriteLine(sb.ToString() + sb2.ToString() + sb3.ToString());
+        }
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/CBDE/MlirExportTest.cs:1347]: this `}` is extra, it doesn't compile

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CbdeHandler.CSharp9.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CbdeHandler.CSharp9.cs:1]: also add the `init` setter, static lambdas, discard params, and some top level statements (and the test should be given `outputKind: OutputKind.ConsoleApplication`)

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs [hunk 6]
```diff
@@ -5170,6 +5170,28 @@ public void Cfg_PropertyPatternClause_Nested()
...
+        {
+            var cfg = Build(@"var x = new Object()");
+
+            VerifyMinimalCfg(cfg);
+            VerifyAllInstructions(cfg.EntryBlock, "new Object()", "x = new Object()");
+        }
+
+        [TestMethod]
+        [TestCategory("CFG")]
+        public void Cfg_New_TargetTyped()
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs:5187]: why did you add a test with only the "target typed new" syntax, and not will the rest of C# 9 syntaxes?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
Nice! I wonder if an integration test can be added

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs:5187]: Mainly because the scope of this PR is CBDE. I just documented the root cause of this known failure to isolate it. I wasn't able to break CFG with other syntaxes.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/MlirExporter.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/MlirExporter.cs:151]: Note from offline discussion: We need a syntax that will cause CFG generation to fail. We have it now, but we will not have it in a long run.
Since it only generates a comment (as the statement above), this should be fine without integration test.

**Comment by pavel-mikula-sonarsource** [comment]
Rebase will be needed after #3733 is merged to deal with the failing UT

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM
